src/WWVSIconSVGOptimizer.class.inc

```diff
@@ -0,0 +1,1605 @@
+ *
+ * Example usage:
+ *
+ * $sanitizer = new WWVSIconSVGOptimizer();
+ * $svg_file_contents = file_get_contents('/path/to/some/svg/file.svg');
+ * $optimized_svg_file_contents = $sanitizer->optimize($svg_file_contents);
+ *
+ * Important websites for reference:
+ *
+ * SVGOMG - https://jakearchibald.github.io/svgomg/
+ * The UI that WW editors use to optimize SVGs before upload and that we are
+ * trying to replicate.
+ *
+ * SVGO - https://github.com/svg/svgo
+ * The JS library that powers the SVGOMG UI and contains all of the logic that
+ * we are trying to replicate.
+ *
+ * BTS2 SVG Configuration Reference -
+ * https://confluence.tools.weightwatchers.com/display/VS/BTS2+SVG+icon+configuration#BTS2SVGiconconfiguration-CleanSVGs
+ * The configuration that editors are required use on SVGOMG to optimize SVGs
+ * and that we are aiming to replicate.
+ *
+ * Helpers:
+ *
+ * Regex Tester - https://regex101.com/
+ * We're processing quite a few regular expressions throughout this library, and
+ * this is a very helpful resource for double checking patterns that you need
+ * confirmation on.
+ *
+ * Process notes:
+ *
+ *  - You can change the test image being processed by this script by changing
+ *    the $test_svg_file variable in wwvs_icon_svg_preprocess_page() in the
+ *    wwvs_icon_svg.module file.
+ *  - Read through the plugins that exist in the SVGO library
+ *    (https://github.com/svg/svgo/tree/master/plugins).
+ *  - Use the example images supplied from the SVGO library in the SVGOMG tool,
+ *    located in the test-svgs directory in the wwvs_icon_svg module directory
+ *    and confirm the changes made by each plugin when it is turned on.
+ *  - Use xdebug to pause (add a breakpoint) before and after processing of each
+ *    plugin to review the changes made. (See $this->svg_file['content'] for the
+ *    output SVG XML.) Make sure you place the ending breakpoint *after* the
+ *    `$this->setXML($xml)` or `$this->setContent($svg)` to review the plugin's
+ *    results.
+ *
+ * Additional notes:
+ *
+ *  - This class depends on the PHP Sanitizer class, which supplies the ability
+ *    to whitelist tags and attributes for SVG cleanliness. This class maintains
+ *    a list of tags and attributes to remove from those whitelists. Any other
+ *    tags or attributes required to be removed must be done using a custom
+ *    plugin method in this class.
+ *  - The base class also provides the ability to minify the file and remove the
+ *    default XML description tag, both of which we opted to do with this class.
+ *    (See $this->setDefaultOptions().) However, the Sanitizer minify option
+ *    does not remove empty text elements, which we are doing in this class.
+ *    (See $this->removeEmptyText().)
+ *
+ */
+
+libraries_load('svg-sanitizer');
+
+module_load_include('inc', 'wwvs_icon_svg', 'includes/WWVSIconSVGOptimizerAttributes.class');
+module_load_include('inc', 'wwvs_icon_svg', 'includes/WWVSIconSVGOptimizerTags.class');
```

Comment:
Drupal code?

Reply:
This is still a work in progress and the Drupal Specific code will be removed.